lib/miq_automation_engine/service_models/miq_ae_service_container_deployment.rb

```diff
@@ -0,0 +1,231 @@
+      pool_id = pool_cmdret[:stdout].split("\n").first.to_s.strip
+      perform_agent_commands(deployment_master, user, ["sudo subscription-manager attach --pool='#{pool_id}'"]) unless pool_id.blank?
+      enabled_repos_cmdret = perform_agent_commands(deployment_master, user, ['sudo subscription-manager repos --disable="*"',
+                                                                              "sudo subscription-manager repos --enable=#{SUBSCRIPTION_REPOS.join(" --enable=")}",
+                                                                              "sudo subscription-manager repos --list-enabled"])
+      enabled_repos = enabled_repos_cmdret[:stdout].split(" ").to_a.map(&:strip).reject(&:empty?)
+      missing_repos = SUBSCRIPTION_REPOS - enabled_repos
+      raise StandardError, "couldn't register to following repos : #{missing_repos.join(',')} with SKU : #{rhsm_sku}" unless missing_repos.empty?
+    end
+
+    def subscribe_cluster(deployment_master, user)
+      perform_agent_commands(deployment_master, user, ["sudo mv ~/#{RHEL_SUBSCRIPTION_FILE_NAME} #{INVENTORIES_PATH}"])
+      run_playbook_command(deployment_master, user, "ansible-playbook #{INVENTORIES_PATH}playbooks/byo/rhel_subscribe.yml -i "\
+                         "#{INVENTORIES_PATH}#{RHEL_SUBSCRIPTION_FILE_NAME} 1> #{ANSIBLE_LOG} 2> #{ANSIBLE_ERROR_LOG} < /dev/null")
+    end
+
+    def provision_tasks
+      automation_task.automation_request.reload
+      return [] unless automation_task.automation_request.options["master_request_ids"]
+      ids = automation_task.miq_request.options["master_request_ids"] + automation_task.miq_request.options["node_request_ids"]
+      MiqRequest.where(:id => ids)
+    end
+
+    def provision_started?
+      provision_tasks.count > 0
+    end
+
+    def nodes_subscription_needed?
+      container_deployment_nodes.count > 1
+    end
+
+    def ssh_user
+      ssh_auth.userid
+    end
+
+    def add_automation_task(task)
+      object_send(:update_attributes!, :automation_task_id => task.id)
+    end
+
+    def customize(options = {})
+      ar_method do
+        @object.customizations = {:agent => {}} if @object.customizations.empty?
+        @object.customizations[:agent].merge!(options)
+        @object.save!
+      end
+    end
+
+    def perform_scp(ip, username, local_path, remote_path)
+      Net::SCP.upload!(ip, username, local_path, remote_path, :ssh => {:key_data => ssh_auth.auth_key})
+    end
+
+    def perform_agent_commands(ip, username, commands = [])
+      @ssh ||= LinuxAdmin::SSH.new(ip, username, ssh_auth.auth_key)
+      @ssh.perform_commands(commands)
+    end
+
+    def provision_vms_status
+      finished = provision_tasks.where(:request_state => 'finished', :status => 'Ok').count == provision_tasks.count
+      fail = provision_tasks.where(:request_state => 'finished', :status => 'Error').count > 0
+      [finished, fail]
+    end
+
+    def find_vm_by_type(tag)
```

Comment:
since we are doing a find by type it would read better if we changed the argument name to be type